tools/ack-discover/ecrpublic.py

```diff
@@ -0,0 +1,136 @@
+# Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may
+# not use this file except in compliance with the License. A copy of the
+# License is located at
+#
+#	 http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+import dataclasses
+import uuid
+
+import boto3
+
+BASE_ECR_URL = "public.ecr.aws/aws-controllers-k8s"
+ECR_PUBLIC_AWS_ACCOUNT_ID = "628432846661"
+ECR_PUBLIC_READER_ROLE_ARN = f"arn:aws:iam::{ECR_PUBLIC_AWS_ACCOUNT_ID}:role/ArtifactReader"
+
+
+@dataclasses.dataclass
+class AWSCredentials:
+    access_key_id: str
+    secret_access_key: str
+    session_token: str
+
+
+def _assume_reader_role(writer):
+    """Returns an access key ID, secret access key, and a security token after
+    assuming an ECR Public readonly role.
+    """
+    writer.debug("[ecr.assume_reader_role] assuming ECR Public ReadOnly role ...")
+    role_session_name = str(uuid.uuid4())
+    sts_client = boto3.client("sts")
+    resp = sts_client.assume_role(
+        RoleArn=ECR_PUBLIC_READER_ROLE_ARN,
+        RoleSessionName=role_session_name,
+        DurationSeconds=60*60,  # 1 hour...
+    )
+    creds = resp['Credentials']
+    return AWSCredentials(
+        access_key_id=creds['AccessKeyId'],
+        secret_access_key=creds['SecretAccessKey'],
+        session_token=creds['SessionToken'],
+    )
+
+
+def get_client(writer):
+    """Returns the ECR Public client object after assuming the ECR Public
+    Reader role.
+    """
+    aws_creds = _assume_reader_role(writer)
```

Comment:
Why do we need to assume a role to read public repositories?

Reply:
> Why do we need to assume a role to read public repositories?

That should probably be a question for the ECR Public team :) You need to have permissions to call the ECR Public `DescribeRepositories` API call...

Reply:
> $ aws ecr-public describe-repositories --repository-name ack-ec2-chart --registry-id 628432846661 --region us-east-1 
An error occurred (AccessDeniedException) when calling the DescribeRepositories operation: User: arn:aws:sts::XXXXXXXXXXXX:role/Admin is not authorized to perform: ecr-public:DescribeRepositories on resource: arn:aws:ecr-public::628432846661:repository/ack-ec2-chart because no resource-based policy allows the ecr-public:DescribeRepositories action

Omg that is an interesting design choice. Wow. Well I know Python has a bunch of really nice screen scraping utilities haha